# SPDX-License-Identifier: Apache-2.0
# Downstream Contextlib2 21.6.0 declaration-only compatibility adjustment.
# Preserve runtime, all original tests/default commands and original allowlist.
# Declare existing GenericAlias subscription and remaining coroutine methods.
--- a/contextlib2/__init__.pyi
+++ b/contextlib2/__init__.pyi
@@ -5,4 +5,6 @@
 from types import TracebackType
+if sys.version_info >= (3, 9):
+    from types import GenericAlias
 from typing import (
     IO,
     Any,
@@ -28,8 +30,15 @@
-AbstractContextManager = ContextManager
-if True:
-    AbstractAsyncContextManager = AsyncContextManager
-
 _T = TypeVar("_T")
 _T_co = TypeVar("_T_co", covariant=True)
+
+class AbstractContextManager(ContextManager[_T_co]):
+    if sys.version_info >= (3, 9):
+        @classmethod
+        def __class_getitem__(cls, item: Any) -> GenericAlias: ...
+
+class AbstractAsyncContextManager(AsyncContextManager[_T_co]):
+    if sys.version_info >= (3, 9):
+        @classmethod
+        def __class_getitem__(cls, item: Any) -> GenericAlias: ...
+
 _T_io = TypeVar("_T_io", bound=Optional[IO[str]])
 _F = TypeVar("_F", bound=Callable[..., Any])
@@ -112,5 +121,5 @@
         def __init__(self) -> None: ...
         def enter_context(self, cm: ContextManager[_T]) -> _T: ...
-        def enter_async_context(self, cm: AsyncContextManager[_T]) -> Awaitable[_T]: ...
+        async def enter_async_context(self, cm: AsyncContextManager[_T]) -> _T: ...
         def push(self, exit: _CM_EF) -> _CM_EF: ...
         def push_async_exit(self, exit: _ACM_EF) -> _ACM_EF: ...
@@ -119,3 +128,3 @@
         def pop_all(self: _S) -> _S: ...
-        def aclose(self) -> Awaitable[None]: ...
+        async def aclose(self) -> None: ...
         async def __aenter__(self: _S) -> _S: ...
@@ -135,3 +144,5 @@
         def __init__(self: nullcontext[_T], enter_result: _T) -> None: ...
         def __enter__(self) -> _T: ...
         def __exit__(self, *exctype: Any) -> bool: ...
+        async def __aenter__(self) -> _T: ...
+        async def __aexit__(self, *excinfo: Any) -> None: ...
